src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -286,20 +290,23 @@ public void GetResult()
 
         public void Dispose()
         {
-            AbortAwaiting();
-
-            // Return all blocks
-            var block = _head;
-            while (block != null)
+            lock (_sync)
             {
-                var returnBlock = block;
-                block = block.Next;
+                AbortAwaiting();
 
-                returnBlock.Pool.Return(returnBlock);
-            }
+                // Return all blocks
+                var block = _head;
+                while (block != null)
+                {
+                    var returnBlock = block;
+                    block = block.Next;
 
-            _head = null;
-            _tail = null;
+                    returnBlock.Pool.Return(returnBlock);
```

Comment:
If the blocks are currently being consumed (i.e. We are between calls to ConsumingStart and ConsumingComplete). I don't think it's safe to return them to the pool yet. I think we need to wait until the following call to ConsumingComplete.